# Fase B — Pipeline ETL
**Proyecto 1 — Bodega de datos, ODS delitos Colombia 2025 (DIJIN)**
Grupo 7 · R.A. 1

Este notebook implementa el pipeline **Extracción → Transformación → Carga** de los 8 reportes
`.xlsx` de la Policía Nacional (DIJIN) hacia la bodega dimensional definida en la Fase A
(`entrega/fase_a_diseño.pdf`).

Es ejecutable de principio a fin con **Restart & Run All**.


In [1]:
import os
from pathlib import Path

import pandas as pd
from dotenv import load_dotenv
from sqlalchemy import create_engine, text
from sqlalchemy.exc import OperationalError

RAIZ_PROYECTO = Path.cwd().parent if Path.cwd().name == "entrega" else Path.cwd()
load_dotenv(RAIZ_PROYECTO / ".env")

CARPETA_DATOS = Path("..", "datos")
DATABASE_URL = os.getenv("DATABASE_URL") or f"sqlite:///{(CARPETA_DATOS / 'bodega_delitos.db').as_posix()}"

print(f"DATABASE_URL configurada: {'sí' if DATABASE_URL else 'NO -- revisa tu .env'}")


DATABASE_URL configurada: sí


## 1. Extracción

Se leen los 8 archivos `.xlsx`. Cada archivo trae el encabezado real en la **fila 11**
(`header=10`) y no incluye una columna de tipo de delito, así que se agrega a partir del
nombre del archivo.

In [2]:
archivos = sorted(CARPETA_DATOS.glob("*.xlsx"))
assert archivos, f"No se encontraron archivos .xlsx en {CARPETA_DATOS.resolve()}"

marcos = []
for archivo in archivos:
    nombre_delito = archivo.stem.removesuffix("2025").strip().replace("_", " ").upper()
    datos = pd.read_excel(archivo, header=10)
    datos.columns = [c.strip() for c in datos.columns]
    # dos archivos usan '*AGRUPA EDAD PERSONA*' en vez de 'AGRUPA_EDAD_PERSONA'
    if "*AGRUPA EDAD PERSONA*" in datos.columns:
        datos = datos.rename(columns={"*AGRUPA EDAD PERSONA*": "AGRUPA_EDAD_PERSONA"})
    datos["TIPO_DELITO"] = nombre_delito
    marcos.append(datos)
    print(f"{archivo.name:45s} -> {datos.shape[0]:>7,} filas, {datos.shape[1]} columnas")

df_crudo = pd.concat(marcos, ignore_index=True)
print(f"\nTotal consolidado (crudo): {len(df_crudo):,} filas")


Homicidio Intencional2025.xlsx                ->  13,727 filas, 9 columnas
Homicidios en accidente de tránsito2025.xlsx  ->   7,536 filas, 9 columnas
Hurto a personas2025.xlsx                     -> 290,891 filas, 9 columnas
Hurto a residencias2025.xlsx                  ->  25,636 filas, 9 columnas
Hurto piratería terrestre2025.xlsx            ->     105 filas, 7 columnas
Secuestro2025.xlsx                            ->     656 filas, 9 columnas
Terrorismo2025.xlsx                           ->     214 filas, 7 columnas
Violencia intrafamiliar2025.xlsx              -> 141,366 filas, 9 columnas

Total consolidado (crudo): 480,131 filas


## 2. Transformación

Limpieza mínima:
- Se eliminan las filas de pie de página de cada Excel (créditos/fuente), identificables porque
  no tienen `CODIGO DANE`.
- Conversión de tipos: `FECHA HECHO` a fecha, `CODIGO DANE` a texto (es un código, no una
  cantidad), `CANTIDAD` a entero.
- Normalización de texto: mayúsculas y sin espacios extra en las columnas categóricas.
- Los valores faltantes de `GENERO` y `AGRUPA_EDAD_PERSONA` (ausentes por completo en
  Terrorismo y Hurto piratería terrestre, y algunos nulos puntuales en el resto) se marcan
  explícitamente como `"NO REPORTADO"` en vez de dejarlos vacíos.

In [3]:
df = df_crudo.dropna(subset=["CODIGO DANE"]).copy()
filas_eliminadas = len(df_crudo) - len(df)
print(f"Filas de pie de página eliminadas: {filas_eliminadas}")

df["FECHA HECHO"] = pd.to_datetime(df["FECHA HECHO"], format="%d/%m/%Y", errors="coerce")
assert df["FECHA HECHO"].isna().sum() == 0, "quedaron fechas sin parsear"

df["CODIGO DANE"] = df["CODIGO DANE"].astype("int64").astype(str)
df["CANTIDAD"] = df["CANTIDAD"].astype(int)

for columna in ["DEPARTAMENTO", "MUNICIPIO", "ARMAS MEDIOS", "GENERO", "AGRUPA_EDAD_PERSONA"]:
    df[columna] = df[columna].fillna("NO REPORTADO").astype(str).str.strip().str.upper()
    df[columna] = df[columna].replace({"": "NO REPORTADO", "NAN": "NO REPORTADO"})

print(f"Filas tras limpieza: {len(df):,}")
df.to_csv(CARPETA_DATOS / "delitos_2025_consolidado.csv", index=False)
df.head(3)


Filas de pie de página eliminadas: 40
Filas tras limpieza: 480,091


,ARMAS MEDIOS,DEPARTAMENTO,MUNICIPIO,FECHA HECHO,GENERO,AGRUPA_EDAD_PERSONA,CODIGO DANE,CANTIDAD,TIPO_DELITO
0,ARMA BLANCA / CORTOPUNZANTE,CUNDINAMARCA,BOGOTÁ D.C. (CT),2025-01-01,MASCULINO,ADULTOS,11001000,1,HOMICIDIO INTENCIONAL
1,ARMA BLANCA / CORTOPUNZANTE,CUNDINAMARCA,BOGOTÁ D.C. (CT),2025-01-01,MASCULINO,ADULTOS,11001000,1,HOMICIDIO INTENCIONAL
2,ARMA BLANCA / CORTOPUNZANTE,CUNDINAMARCA,BOGOTÁ D.C. (CT),2025-01-01,MASCULINO,ADULTOS,11001000,1,HOMICIDIO INTENCIONAL


### 2.1 Construcción de dimensiones (claves subrogadas)

Se construyen las 5 tablas de dimensión. Cada una recibe una clave subrogada entera
secuencial, y se valida que sea única antes de continuar.

In [4]:
# dim_tiempo
fechas = (
    df[["FECHA HECHO"]]
    .drop_duplicates()
    .reset_index(drop=True)
    .rename(columns={"FECHA HECHO": "fecha"})
)
fechas["fecha_id"] = fechas["fecha"].dt.strftime("%Y%m%d").astype(int)
fechas["anio"] = fechas["fecha"].dt.year
fechas["trimestre"] = fechas["fecha"].dt.quarter
fechas["mes"] = fechas["fecha"].dt.month
fechas["nombre_mes"] = fechas["fecha"].dt.month_name()
fechas["dia"] = fechas["fecha"].dt.day
fechas["dia_semana"] = fechas["fecha"].dt.day_name()
fechas["es_fin_de_semana"] = fechas["fecha"].dt.dayofweek >= 5
dim_tiempo = fechas[
    ["fecha_id", "fecha", "anio", "trimestre", "mes", "nombre_mes", "dia", "dia_semana", "es_fin_de_semana"]
]
assert dim_tiempo["fecha_id"].is_unique
print(f"dim_tiempo: {dim_tiempo.shape}")


dim_tiempo: (365, 9)


In [5]:
# dim_geografia (jerarquía departamento -> municipio, con código DANE)
dim_geografia = (
    df[["CODIGO DANE", "DEPARTAMENTO", "MUNICIPIO"]]
    .drop_duplicates()
    .reset_index(drop=True)
)
dim_geografia.insert(0, "geo_id", dim_geografia.index + 1)
dim_geografia = dim_geografia.rename(
    columns={"CODIGO DANE": "codigo_dane", "DEPARTAMENTO": "departamento", "MUNICIPIO": "municipio"}
)
assert dim_geografia["geo_id"].is_unique
print(f"dim_geografia: {dim_geografia.shape}")


dim_geografia: (1500, 4)


In [6]:
# dim_tipo_delito
dim_tipo_delito = df[["TIPO_DELITO"]].drop_duplicates().reset_index(drop=True)
dim_tipo_delito.insert(0, "delito_id", dim_tipo_delito.index + 1)
dim_tipo_delito = dim_tipo_delito.rename(columns={"TIPO_DELITO": "tipo_delito"})
assert dim_tipo_delito["delito_id"].is_unique
print(f"dim_tipo_delito: {dim_tipo_delito.shape}")

# dim_arma_medio
dim_arma_medio = df[["ARMAS MEDIOS"]].drop_duplicates().reset_index(drop=True)
dim_arma_medio.insert(0, "arma_id", dim_arma_medio.index + 1)
dim_arma_medio = dim_arma_medio.rename(columns={"ARMAS MEDIOS": "arma_medio"})
assert dim_arma_medio["arma_id"].is_unique
print(f"dim_arma_medio: {dim_arma_medio.shape}")

# dim_victima
dim_victima = df[["GENERO", "AGRUPA_EDAD_PERSONA"]].drop_duplicates().reset_index(drop=True)
dim_victima.insert(0, "victima_id", dim_victima.index + 1)
dim_victima = dim_victima.rename(columns={"GENERO": "genero", "AGRUPA_EDAD_PERSONA": "grupo_edad"})
assert dim_victima["victima_id"].is_unique
print(f"dim_victima: {dim_victima.shape}")


dim_tipo_delito: (8, 2)
dim_arma_medio: (23, 2)
dim_victima: (10, 3)


### 2.2 Construcción de la tabla de hechos

`hecho_delito` tiene grano **"una fila = un caso delictivo registrado"** (el más fino que
entrega la fuente). Se obtiene uniendo `df` con las 5 dimensiones y quedándose solo con las
llaves subrogadas y la medida `cantidad`. Se valida que el join no pierda ni duplique filas y
que no queden llaves foráneas nulas.

In [7]:
h = df.copy()
h["fecha_id"] = h["FECHA HECHO"].dt.strftime("%Y%m%d").astype(int)

h = h.merge(
    dim_geografia,
    left_on=["CODIGO DANE", "DEPARTAMENTO", "MUNICIPIO"],
    right_on=["codigo_dane", "departamento", "municipio"],
    how="left",
)
h = h.merge(dim_tipo_delito, left_on="TIPO_DELITO", right_on="tipo_delito", how="left")
h = h.merge(dim_arma_medio, left_on="ARMAS MEDIOS", right_on="arma_medio", how="left")
h = h.merge(dim_victima, left_on=["GENERO", "AGRUPA_EDAD_PERSONA"], right_on=["genero", "grupo_edad"], how="left")

hecho_delito = h[["fecha_id", "geo_id", "delito_id", "arma_id", "victima_id", "CANTIDAD"]].rename(
    columns={"CANTIDAD": "cantidad"}
)
hecho_delito.insert(0, "hecho_id", hecho_delito.index + 1)

assert len(hecho_delito) == len(df), "el join cambió el número de filas -- revisar duplicados en dimensiones"
fk_nulos = hecho_delito[["fecha_id", "geo_id", "delito_id", "arma_id", "victima_id"]].isna().sum()
assert fk_nulos.sum() == 0, f"quedaron llaves foráneas nulas:\n{fk_nulos}"

print(f"hecho_delito: {hecho_delito.shape} -- {len(hecho_delito):,} filas (igual a df: {len(df):,})")
hecho_delito.head(3)


hecho_delito: (480091, 7) -- 480,091 filas (igual a df: 480,091)


,hecho_id,fecha_id,geo_id,delito_id,arma_id,victima_id,cantidad
0,1,20250101,1,1,1,1,1
1,2,20250101,1,1,1,1,1
2,3,20250101,1,1,1,1,1


## 3. Carga a la bodega

Se escribe a la base de datos indicada por `DATABASE_URL` (variable de entorno en `.env`)
usando SQLAlchemy. El código es el mismo sin importar el motor: para desarrollo/pruebas se usa
SQLite (`sqlite:///datos/bodega_delitos.db`) y para producción PostgreSQL
(`postgresql+psycopg2://usuario:clave@host:5432/bodega_delitos`) — **basta con cambiar la URL
en `.env`, no hay que tocar código.**

Se maneja el error de conexión con un mensaje claro en vez de dejar que truene con una traza
larga.

In [8]:
tablas = {
    "dim_tiempo": dim_tiempo,
    "dim_geografia": dim_geografia,
    "dim_tipo_delito": dim_tipo_delito,
    "dim_arma_medio": dim_arma_medio,
    "dim_victima": dim_victima,
    "hecho_delito": hecho_delito,
}

try:
    engine = create_engine(DATABASE_URL)
    with engine.connect() as conexion:
        for nombre_tabla, tabla in tablas.items():
            tabla.to_sql(nombre_tabla, conexion, if_exists="replace", index=False)
        conexion.commit()
    print("Carga completada correctamente en:", engine.url.render_as_string(hide_password=True))
except OperationalError as error:
    raise SystemExit(
        "No se pudo conectar a la base de datos. Verifica que DATABASE_URL en tu .env "
        "sea correcta y que el servidor esté disponible.\n"
        f"Detalle: {error}"
    )


Carga completada correctamente en: sqlite:///../datos/bodega_delitos.db


## 4. Validación de la carga

Se cuentan las filas de cada tabla ya en la base de datos para confirmar que la carga fue
completa.

In [9]:
with engine.connect() as conexion:
    for nombre_tabla in tablas:
        n = conexion.execute(text(f"SELECT COUNT(*) FROM {nombre_tabla}")).scalar()
        print(f"{nombre_tabla:20s} {n:>10,} filas")


dim_tiempo                  365 filas
dim_geografia             1,500 filas
dim_tipo_delito               8 filas
dim_arma_medio               23 filas
dim_victima                  10 filas
hecho_delito            480,091 filas


## Resumen

- **480,091** casos delictivos consolidados desde los 8 reportes DIJIN 2025.
- Bodega dimensional cargada: 1 tabla de hechos (`hecho_delito`) + 5 dimensiones
  (`dim_tiempo`, `dim_geografia`, `dim_tipo_delito`, `dim_arma_medio`, `dim_victima`).
- 0 filas perdidas, 0 llaves foráneas nulas.
- Listo para las consultas analíticas de la Fase C.